### PR TITLE
Nim ping test

### PR DESCRIPTION
**Pull request by Menduist**
Should work fine now! No metrics though

Old message:

---------
Works fine against itself, could use a hand to setup the interop tests themselves

I'm also not sure what format the pubsub payload should have, because it's being escaped
If I send {"publish": {"topic": "whatev", "payload": {"test": "whatev"}}
I receive something like "{\"test\": \"whatev\"}"

So I sent it as a string, which may fail with other implementations of the SDK

Also missing metrics since I don't support them in the SDK (yet)

### CODE REVIEW

**Comment by laurentsenta** [comment]
Awesome, thanks for sharing,

If you want to give this a shot:
you could create a `all-interop-latest.toml`, copy the content of https://github.com/libp2p/test-plans/blob/master/ping/_compositions/go-rust-interop-latest.toml and add a `[groups]` section with the nim builder. The section should look something like: https://github.com/libp2p/test-plans/blob/b97b36aa2754c77795635facbf96c8504bdd9135/ping/_compositions/go-rust-interop-latest.toml#L18-L29 (ignore templating for now).

On the payload:
that's how the sync service works at the moment (on my todo is to describe the current, implicit, spec).
You get the JSON back, encoded as a string.

Example in rust:
https://github.com/testground/sdk-rust/blob/2b85e4c9578037bd8704de40b7f179bf7fcc8649/src/responses.rs#L110

**Comment by Menduist** [comment]
It works!

Had to fix a few things (the string thing, and the nim SDK doesn't signal `initialized_global` which is apparently required for the other SDKs, so I did it manually here), but seems to work fine now :)

**Comment by laurentsenta** [comment]
Nice!

If all you need is a git commit, recommendation:
create a `nim.toml` which specifies multiple groups with an `Id` and a `GitRef` value.

Check the `rust` part, I guess it would look something like:

```toml
{{ with (load_resource "./nim.toml") }}
  {{ with (index .groups 0) }} // get the latest
    [[groups]]
    id = "nim-latest-{{ .Id }}"
    instances = { count = 1 }
    builder = "docker:generic"

    [groups.build_config]
      path = "./rust/"

    [groups.build_config.build_args]
      GIT_REF = '{{ .GitRef }}'
  {{ end }}
{{ end }}
```

**Comment by Menduist** [comment]
Done, + an attempt at adding to CI

**Comment by Menduist** [comment]
Workflow ran successfully on the fork: https://github.com/status-im/libp2p-test-plans/actions/runs/3427378367
So I would say this is ready for review

File: .github/workflows/ping-interop-all.yml
Comment (mxinden) [inline .github/workflows/ping-interop-all.yml:1]: Instead of introducing a new workflow file, how about including nim in `ping-interop.yml`. I don't see a reason why we should treat Go or Rust special.

Maybe @MarcoPolo has opinions here.

**Review by mxinden — COMMENTED** [review]
:rocket: happy to see this happening.

File: .github/workflows/ping-interop-all.yml
Comment (Menduist) [reply .github/workflows/ping-interop-all.yml:1]: `ping-interop.yml` will call the `go-rust-interop.toml` composition
So it could just call the `all-interop.toml` instead, but then the `go-rust-interop.toml` will never be called, and maybe it should be removed?
We would just have:
- `[language].toml`: list of versions of that implem
- `[language]-cross-versions.toml`: test every version of that implem
- `all-interop-latest.toml`: test last version of every implem
- `all-interop.toml`: test every version of every implem

Comment (Menduist) [reply .github/workflows/ping-interop-all.yml]: Did this in my latest commit, we can revert if not happy

Comment (mxinden) [reply .github/workflows/ping-interop-all.yml]: > We would just have:
> 
>     * `[language].toml`: list of versions of that implem
> 
>     * `[language]-cross-versions.toml`: test every version of that implem
> 
>     * `all-interop-latest.toml`: test last version of every implem
> 
>     * `all-interop.toml`: test every version of every implem

On a high level I am in favor of this. I will take a detailed look now.

File: .github/workflows/ping-interop-latest.yml
Comment (mxinden) [inline .github/workflows/ping-interop-latest.yml:23]: Nit pick, while we are at it, how about being consistent with the naming on `ping-interop.yml`?
```suggestion
name: libp2p ping - go, rust and nim test (latest) with testground.
```

**Review by mxinden — COMMENTED** [review]
Very excited for us to have interoperability tests beyond Go, JS and Rust.

The proposed changes here look good to me.

I suggest waiting for another review from @MarcoPolo (as self-requested).

In addition, I wonder how we should coordinate this with https://github.com/libp2p/test-plans/issues/53, i.e. how we should prioritize this and the refactoring of our test matrix generation. As far as I can tell, it would be fine to proceed here before https://github.com/libp2p/test-plans/issues/53. What do you think @jxs and @MarcoPolo?

**Comment by mxinden** [comment]
>  are the other implementation caching their dependencies properly to be reused by multiple images?

Would you mind merging `master` here @Menduist? https://github.com/libp2p/test-plans/pull/78 might resolve the issues you are seeing.

**Comment by mxinden** [comment]
> I suggest waiting for another review from @MarcoPolo (as self-requested).

Friendly ping @MarcoPolo. Any objections moving forward here?

**Comment by mxinden** [comment]
Triggered the tests. All green @Menduist :tada:

**Review by MarcoPolo — APPROVED** [review]

**Comment by MarcoPolo** [comment]
Looks great! @Menduist I'll ping you for some help in getting the multidimensional testing in the near future. Maybe we can pair on it? Might be fun to pair on some Nim 👑 👑 👑

**Comment by Menduist** [comment]
Sure! Thanks for the merge 🎉

**Pull request merged**
